### ISSUE TITLE
Type mismatch error with Linear layer

### ISSUE THREAD

**Issue by jean-airoldie** [LaurentMazare/tch-rs#100]
I get an errror: `Expected object of scalar type Double but got scalar type Float for argument`
caused by:
https://github.com/LaurentMazare/tch-rs/blob/c6888ba746d7123939bf8f092c06b35a16319b7c/src/nn/linear.rs#L51
when passing a `Kind::Double` to the `forward` method of a `Linear` layer. Seems like there is a type mismatch between the weights and biases.

You can find the full backtrace [here](https://gist.github.com/jean-airoldie/69c4a9c2e5c717bd59dc4dd2b6fe47ec).

Here is the minimal reproducible code:
```rust
use tch::{nn, nn::Module, nn::OptimizerConfig, Device, Tensor, Kind, Reduction};

const INPUT_DIM: i64 = 784;
const BATCH_SIZE: i64 = 1;
const HIDDEN_NODES: i64 = 128;
const LABELS: i64 = 10;

fn net(vs: &nn::Path) -> impl Module {
    nn::seq()
        .add(nn::linear(
            vs / "layer1",
            INPUT_DIM,
            HIDDEN_NODES,
            Default::default(),
        ))
        .add(nn::linear(vs, HIDDEN_NODES, LABELS, Default::default()))
}

pub fn run() -> failure::Fallible<()> {
    let vs = nn::VarStore::new(Device::Cpu);
    let net = net(&vs.root());
    let opt = nn::Adam::default().build(&vs, 1e-3)?;
    for epoch in 1..200 {
        let features = Tensor::rand(&[BATCH_SIZE, INPUT_DIM], (Kind::Double, Device::Cpu));
        let labels = Tensor::rand(&[BATCH_SIZE, LABELS], (Kind::Double, Device::Cpu));

        let predictions = net.forward(&features); // This panics.

        let loss = predictions.mse_loss(&labels, Reduction::Mean);
        opt.backward_step(&loss);
        println!(
            "epoch: {:4} train loss: {:8.5}",
            epoch,
            f64::from(&loss),
        );
    }
    Ok(())
}
```

**Comment by LaurentMazare** [comment]
Yes that's actually expected: all the layers in `nn` use single precision float and not double precision ones. It feels like the typical use case for neural networks - although supporting half-precision float would be nice.
I imagine you already know about this but a quick workaround for your use case is to use `Kind::Float` for `features` and `labels`.
We could add to the parameters the kind of weights to actually generate if you think that would be helpful, although I'm not a huge fan of the idea as this is currently not reflected in the type system.

**Comment by vegapit** [comment]
I remember seeing benchmarks showing the improved performance of using `f32` versus `f64`. I have since assumed that it was the reason why low-level math libraries seem to use `Float` instead of `Double`.

**Comment by jean-airoldie** [comment]
In my use case, I can't really justify using a f64, since all my inputs are already normalized. But I wonder if there could be valid use cases for `f64` or `f16`.

I think that ideally we would want models to be typed to prevent those kind of errors, but that might be too high level for this library.

Anyway, what do you think of adding some debug assertions so that we get better error messages? The errors returned by the `c++` lib are mostly garbage.

**Comment by jean-airoldie** [comment]
@vegapit In my experience, there are no significant performance improvement at the single operation level, but the fact that you can fit 2X more in memory is obviously better for the cache. Also, if your bottleneck is a SSE operations, you effectively double the throughput.

**Comment by LaurentMazare** [comment]
The performance implications are indeed quite important, it's actually why half-precision floats are gaining momentum as on a gpu they may result on an almost 2x speedup. Double-precision floats are also notoriously slow on non-tesla gpus.
It would certainly be worth having proper error messages rather than relying on PyTorch to produce the error message. Feel free to make a PR for this if you want, otherwise I'll take a stab at it when I get a chance.

**Comment by jean-airoldie** [comment]
Yeah, I'll make a PR.

But more generally, do you think that preventing this kind of error using the type system is out of the scope of this library?

**Comment by LaurentMazare** [comment]
> But more generally, do you think that preventing this kind of error using the type system is out of the scope of this library?

It's a bit unclear. It's out of scope of the `Tensor` structure as this part is automatically generated from some ops descriptions and these descriptions don't include enough type information to get this to be type safe.
For the `nn` bits, having more type safety would make more sense but that's likely to be a huge change so I'm not sure it's worth doing so: I guess it would be more a matter of having another namespace `nn_typed` or even another crate on top this to handle it.
Happy to hear about other people thoughts too.

**Comment by jean-airoldie** [comment]
I think a higher level wrapper type `Tensor<D,T>` where D is the dimension and T the underlying type would be ideal since it would basically prevent all possible runtime dynamic errors. But like you said the amount of work would be quite significant.

**Comment by jean-airoldie** [comment]
> We could add to the parameters the kind of weights to actually generate if you think that would be helpful, although I'm not a huge fan of the idea as this is currently not reflected in the type system.

I think it would make sense to allow the kind of weights since this library is a binding library. While I agree this should ideally be reflected in the type system, I also think the library should not restrain use cases.

**Comment by MichaelBurge** [comment]
I ran into this issue too. Tensor operations like `MulAssign` are only implemented for `f64` and not `f32`, and it is somewhat convenient to use the same float type for everything.
